Tag API and several random issues fixed

UmashriR opened the pull request.

Umashri committed:
Tag API and several random issues fixed

vijayakrishnap commented on ZCRMiOS/Classes/EntityAPIHandler.swift:
```diff
@@ -752,7 +747,10 @@ internal class EntityAPIHandler : CommonAPIHandler
             }
         }
-        tagNamesString.removeLast()
+        if tagNamesString.startIndex != tagNamesString.endIndex
```
What is this check?

UmashriR replied:
If the tagNamesString is empty, there occurs a run time error

vijayakrishnap commented on ZCRMiOS/Classes/EntityAPIHandler.swift:
```diff
@@ -773,15 +771,19 @@ internal class EntityAPIHandler : CommonAPIHandler
     }
     
-    internal func removeTags( tags : [ZCRMTag], completion : @escaping( ZCRMTag?, APIResponse?, Error? ) -> () )
+    internal func removeTags( tags : [ZCRMTag], completion : @escaping( [ZCRMTag]?, APIResponse?, Error? ) -> () )
```
For this method, we can change the completion handler as (APIResponse, Error), since the list of tags wouldn't have any scope after removing them. This response should be like delete APIs responses.

vijayakrishnap commented on ZCRMiOS/Classes/MetaDataAPIHandler.swift:
```diff
@@ -122,8 +122,9 @@ internal class MetaDataAPIHandler : CommonAPIHandler
             {
-                dollarProperty.replacingOccurrences(of: "$", with: "")
-                properties.append(dollarProperty)
+                var property = dollarProperty
```
replacing $ not needed?

vijayakrishnap commented on ZCRMiOS/Classes/MassEntityAPIHandler.swift:
```diff
@@ -462,7 +465,10 @@ internal class MassEntityAPIHandler : CommonAPIHandler
             }
         }
-        tagNamesString.removeLast()
+        if tagNamesString.startIndex != tagNamesString.endIndex
```
Whats this check for?

UmashriR replied:
If the tagNamesString is empty, there occurs a run time error

vijayakrishnap commented on ZCRMiOS/Classes/ZCRMModule.swift:
```diff
@@ -1019,31 +1019,24 @@ public class ZCRMModule : ZCRMEntity
     }
     
-    public func deleteTag( tag : ZCRMTag, completion : @escaping ( APIResponse?, Error? ) -> () )
+    public func addTags( recordIds : [Int64], tags : [ZCRMTag], completion : @escaping( [ZCRMRecord]?, BulkAPIResponse?, Error? ) -> () )
```
Why is addTags returning [ZCRMRecord] ? shouldn't it be [ZCRMTag] ?

UmashriR replied:
Discussed

vijayakrishnap commented on ZCRMiOS/Classes/ZCRMModule.swift:
```diff
@@ -1019,31 +1019,24 @@ public class ZCRMModule : ZCRMEntity
     }
     
-    public func removeTags( recordIds: [Int64], tags : [ZCRMTag], completion : @escaping( [ZCRMTag]?, BulkAPIResponse?, Error? ) -> () )
+    public func removeTags( recordIds: [Int64], tags : [ZCRMTag], completion : @escaping( [ZCRMRecord]?, BulkAPIResponse?, Error? ) -> () )
```
For this also, let the completion handler arguments be (BulkAPIResponse, Error)

vijayakrishnap commented on ZCRMiOS/Classes/ZCRMRecord.swift:
```diff
@@ -288,6 +289,16 @@ public class ZCRMRecord : ZCRMEntity
         return self.lineItems
     }
     
+    internal func setTags( tag : ZCRMTag )
```
since its adding only one tag, the method name should be addTag

vijayakrishnap commented on ZCRMiOS/Classes/ZCRMRecord.swift:
```diff
@@ -754,24 +758,24 @@ public class ZCRMRecord : ZCRMEntity
     }
     
-    public func addTags( tags : [ZCRMTag], completion : @escaping( ZCRMTag?, APIResponse?, Error? ) -> () )
+    public func addTags( tags : [ZCRMTag], completion : @escaping( [ZCRMTag]?, APIResponse?, Error? ) -> () )
```
I think for this also we can give completion handler arguments as (APIResponse, Error) and remove [ZCRMTag] since these tag objects are not newly created objects.

UmashriR replied:
Discussed

Umashri committed:
Changes specified are done

vijayakrishnap commented on ZCRMiOS/Classes/MassEntityAPIHandler.swift:
```diff
@@ -545,7 +544,7 @@ internal class MassEntityAPIHandler : CommonAPIHandler
             }
         }
-        if tagNamesString.startIndex != tagNamesString.endIndex
+        if tagNamesString.count != 0 && tagNamesString.last == ","
```
Write this method as an extension to String class or a method in CommonUtil.

vijayakrishnap commented on ZCRMiOS/Classes/ZCRMRecord.swift:
```diff
@@ -289,7 +289,7 @@ public class ZCRMRecord : ZCRMEntity
     }
     
-    internal func setTags( tag : ZCRMTag )
+    internal func addTags( tag : ZCRMTag )
```
Only one tag is added. So its "addTag".

vijayakrishnap commented on ZCRMiOS/Classes/MetaDataAPIHandler.swift:
```diff
@@ -118,15 +118,7 @@ internal class MetaDataAPIHandler : CommonAPIHandler
-                properties.append(property)
-            }
-            module.setProperties(properties: properties)
+            module.setProperties(properties: moduleDetails.optArray(key: "$properties") as! [String])
```
Is this change new?

UmashriR replied:
As you said, replacing $ in $properties is not needed, code for replacing is removed

Umashri committed:
Requested changes done.
ModifiedSince, page and perPage added to getDeletedRecords in MassEntityAPIHandler.
